agents/dqn.py

```diff
@@ -1,24 +1,132 @@
 
-    def __init__(self, parts):
+    return argmax_next_neighbor * mask + random_next_neighbor * reverse_mask
+
+
+class DQNAgent(TorchAgent, config_name='dqn'):
+    def __init__(
+            self,
+            current_node_idx_prefix: str,
+            destination_node_idx_prefix: str,
+            neighbors_node_ids_prefix: str,
+            node_idx: int,
+            q_network: BaseQNetwork,
+            discount_factor: float,
+            research_prob: float,
+            bag_trajectory_memory: BaseBagTrajectoryMemory,
+            sample_size: int,
+    ):
+        assert 0 < discount_factor < 1, 'Incorrect `discount_factor` choice'
+        assert 0 < research_prob < 1, 'Incorrect `discount_factor` choice'
         super().__init__()
-        self._parts = parts
+
+        self._current_node_idx_prefix = current_node_idx_prefix
+        self._destination_node_idx_prefix = destination_node_idx_prefix
+        self._neighbors_node_ids_prefix = neighbors_node_ids_prefix
+
+        self._node_idx = node_idx
+        self._q_network = q_network
+        self._discount_factor = discount_factor
+        self._research_prob = research_prob
+        self._bag_trajectory_memory = bag_trajectory_memory
+        self._sample_size = sample_size
 
     @classmethod
     def create_from_config(cls, config):
         return cls(
-            parts=[
-                BaseEncoder.create_from_config(part)
-                for part in config['parts']
-            ]
+            current_node_idx_prefix=config['current_node_idx_prefix'],
+            destination_node_idx_prefix=config['destination_node_idx_prefix'],
+            neighbors_node_ids_prefix=config['neighbors_node_ids_prefix'],
+            q_network=BaseEncoder.create_from_config(config['q_network']),
+            discount_factor=config.get('discount_factor', 0.99),
+            research_prob=config.get('research_prob', 0.1),
+            bag_trajectory_memory=BaseBagTrajectoryMemory.create_from_config(config['path_memory'])
+            if 'path_memory' in config else None,
+            node_idx=config.get('node_idx', None),
+            sample_size=config.get('sample_size', 100)
+        )
+
+    def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        # Shape: [batch_size]
+        current_node_idx = inputs[self._current_node_idx_prefix]
+        # Shape: [batch_size]
+        destination_node_idx = inputs[self._destination_node_idx_prefix]
+        # TensorWithMask
+        neighbor_node_ids = inputs[self._neighbors_node_ids_prefix]
+
+        next_neighbor_q = self._q_network(
+            current_node_idx=current_node_idx,
+            neighbor_node_ids=neighbor_node_ids,
+            destination_node_idx=destination_node_idx
         )
 
-    def forward(self, inputs):
-        for part in self._parts:
-            inputs = part(inputs)
-        return inputs
+        next_neighbor_ids = neighbor_node_ids.flatten_values[torch.argmax(next_neighbor_q, dim=1)]
```

Comment:
What if `torch.argmax` will return the index of the padded neighbor? Is it possible? Can you please reply to it? Also, can you please write info about shapes in order to simplify code review and further debugging?

Reply:
no, q_network returns tensor with -inf in it